File: MIDTERM-LAB-1/data-quality-report.js
const fs = require("fs");
const readline = require("readline");

// ── Config ──────────────────────────────────────────────────────────────────
const SALES_COLUMNS = ["total_sales", "na_sales", "jp_sales", "pal_sales", "other_sales"];
const DATE_FORMATS = [/^\d{4}-\d{2}-\d{2}$/, /^\d{2}\/\d{2}\/\d{4}$/, /^\d{2}-\d{2}-\d{4}$/, /^\w+ \d{2}, \d{4}$/];

const R = "\x1b[0m", GREEN = "\x1b[32m", YELLOW = "\x1b[33m", RED = "\x1b[31m", CYAN = "\x1b[36m", BOLD = "\x1b[1m";

// ── 7. Modular Design (Helper Functions) ─────────────────────────────────────
function section(title) {
    console.log(`\n${BOLD}  ▸ ${title}${R}\n`);
}

function isValidDate(value) {
    if (!value || value.trim() === "") return false;
    const d = new Date(value.trim());
    return !isNaN(d.getTime()) || DATE_FORMATS.some(re => re.test(value.trim()));
}

function parseCsv(content) {
    // 8. Try-Catch for Format Validation
    try {
        const lines = content.split(/\r?\n/).filter(l => l.trim() !== "");
        if (lines.length === 0) throw new Error("File is empty.");

        const parse = line => {
            const fields = [];
            let cur = "", inQ = false;
            for (let i = 0; i < line.length; i++) {
                const c = line[i];
                if (c === '"') {
                    if (inQ && line[i + 1] === '"') { cur += '"'; i++; }
                    else inQ = !inQ;
                } else if (c === "," && !inQ) { fields.push(cur); cur = ""; }
                else cur += c;
            }
            fields.push(cur);
            return fields;
        };

        const headers = parse(lines[0]).map(h => h.trim().toLowerCase());
        const rows = lines.slice(1).map(l => {
            const vals = parse(l);
            const obj = {};
            headers.forEach((h, i) => { obj[h] = vals[i] !== undefined ? vals[i].trim() : ""; });
            return obj;
        });
        return { headers, rows };
    } catch (e) {
        throw new Error("Invalid CSV format: " + e.message);
    }
}

// ── 1 & 3. Input Loop & Validation ───────────────────────────────────────────
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

function askForFile() {
    rl.question(`${BOLD}Enter dataset file path:${R} `, (filePath) => {
        try {
            // 2. Validate existence, readability, and format
            if (!fs.existsSync(filePath)) {
                throw new Error("File does not exist.");
            }
            
            // Check readability
            fs.accessSync(filePath, fs.constants.R_OK);

            const content = fs.readFileSync(filePath, "utf8");
            const { headers, rows } = parseCsv(content);

            runReport(filePath, headers, rows);
            rl.close();
        } catch (err) {
            // 3. Display appropriate error and loop
            console.log(`${RED}[ERROR] ${err.message}${R}`);
            askForFile(); 
        }
    });
}

// ── 9. Formatted Output ──────────────────────────────────────────────────────
function runReport(csvFile, headers, rows) {
    const n = rows.length;
    console.log(`\n${BOLD}${CYAN}╔══════════════════════════════════╗`);
    console.log("║    DATA QUALITY REPORT           ║");
    console.log(`╚══════════════════════════════════╝${R}`);
    console.log(`  File: ${csvFile}`);
    console.log(`  Rows: ${n}  |  Columns: ${headers.length}`);

    // Section 1: Missing Values
    section("1. MISSING VALUES");
    console.log(`  ${"Column".padEnd(30)}  ${"Missing".padStart(8)}  ${"% Null".padStart(6)}`);
    console.log("  " + "─".repeat(50));
    headers.forEach(col => {
        const count = rows.filter(r => !r[col]).length;
        const pct = n > 0 ? ((count / n) * 100).toFixed(1) : "0.0";
        const color = count === 0 ? GREEN : parseFloat(pct) > 20 ? RED : YELLOW;
        console.log(`  ${col.padEnd(30)}  ${color}${String(count).padStart(8)}  ${String(pct + "%").padStart(6)}${R}`);
    });

    // Section 2: Negative Sales
    section("2. NEGATIVE SALES CHECK");
    SALES_COLUMNS.forEach(col => {
        if (!headers.includes(col)) return;
        const neg = rows.filter(r => parseFloat(r[col]) < 0).length;
        console.log(`  ${col.padEnd(20)}  ${neg === 0 ? GREEN : RED}${String(neg).padStart(12)}  ${neg === 0 ? "✓ Clean" : "✗ Issues"}${R}`);
    });

    console.log(`\n${BOLD}${CYAN}${"═".repeat(38)}\n  END OF REPORT\n${"═".repeat(38)}${R}`);
}

// Start the program
askForFile();

File: MIDTERM-LAB-1/JAVA/DataQualityReport.java
import java.io.*;
import java.nio.file.*;
import java.text.*;
import java.util.*;

public class DataQualityReport {

    private static final List<String> SALES_COLUMNS = Arrays.asList(
        "total_sales", "na_sales", "jp_sales", "pal_sales", "other_sales"
    );

    private static final String[] DATE_FORMATS = {
        "yyyy-MM-dd", "MM/dd/yyyy", "dd/MM/yyyy", "yyyy/MM/dd"
    };

    private static final String RESET  = "\u001B[0m";
    private static final String GREEN  = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String RED    = "\u001B[31m";
    private static final String CYAN   = "\u001B[36m";
    private static final String BOLD   = "\u001B[1m";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Path filePath = null;
        List<DataRecord> records = new ArrayList<>();
        String[] headers = null;

        while (true) {
            System.out.println(BOLD + "Enter the FULL FILE PATH of the CSV dataset:" + RESET);
            System.out.print("> ");
            String inputPath = scanner.nextLine().replace("\"", "");
            filePath = Paths.get(inputPath);

            if (!Files.exists(filePath)) {
                System.out.println(RED + "[ERROR] File does not exist." + RESET);
            } else if (!Files.isReadable(filePath)) {
                System.out.println(RED + "[ERROR] File is not readable." + RESET);
            } else if (!inputPath.toLowerCase().endsWith(".csv")) {
                System.out.println(RED + "[ERROR] Must be a .csv file." + RESET);
            } else {
                try {
                    List<DataRecord> tempRecords = loadCSV(filePath);
                    if (tempRecords.isEmpty()) {
                        System.out.println(RED + "[ERROR] The file is empty." + RESET);
                        continue;
                    }
                    headers = tempRecords.get(0).getFields();
                    for (int i = 1; i < tempRecords.size(); i++) {
                        records.add(tempRecords.get(i));
                    }
                    break; 
                } catch (IOException e) {
                    System.out.println(RED + "[ERROR] Failed to read file: " + e.getMessage() + RESET);
                }
            }
        }

        runReport(filePath.getFileName().toString(), headers, records);
        scanner.close(); 
    }

    private static List<DataRecord> loadCSV(Path path) throws IOException {
        List<DataRecord> records = new ArrayList<>();
        try (BufferedReader br = Files.newBufferedReader(path)) {
            String line;
            while ((line = br.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    records.add(new DataRecord(parseCsvLine(line)));
                }
            }
        }
        return records;
    }

    private static void runReport(String fileName, String[] headers, List<DataRecord> records) {
        int totalRows = records.size();
        Map<String, Integer> colMap = new LinkedHashMap<>();
        for (int i = 0; i < headers.length; i++) {
            colMap.put(headers[i].trim().toLowerCase(), i);
        }

        System.out.println(BOLD + CYAN + "\n╔══════════════════════════════════╗");
        System.out.println("║     DATA QUALITY REPORT          ║");
        System.out.println("╚══════════════════════════════════╝" + RESET);
        System.out.println(" File: " + fileName + " | Rows: " + totalRows + "\n");

        System.out.println(BOLD + " ▸ 1. MISSING VALUES" + RESET);
        for (String col : colMap.keySet()) {
            int idx = colMap.get(col);
            long missing = 0;
            for(DataRecord r : records) { if(r.getField(idx).isEmpty()) missing++; }
            double pct = (missing * 100.0 / totalRows);
            String color = missing == 0 ? GREEN : (pct > 20 ? RED : YELLOW);
            System.out.printf(" %-30s %s%8d %7.1f%%%s\n", col, color, missing, pct, RESET);
        }

        System.out.println("\n" + BOLD + " ▸ 2. NEGATIVE SALES CHECK" + RESET);
        for (String saleCol : SALES_COLUMNS) {
            if (colMap.containsKey(saleCol)) {
                int idx = colMap.get(saleCol);
                long negatives = 0;
                for(DataRecord r : records) {
                    try { if(Double.parseDouble(r.getField(idx)) < 0) negatives++; } catch(Exception e){}
                }
                System.out.printf(" %-20s %8d  %s%s\n", saleCol, negatives, (negatives == 0 ? GREEN + "✓ Clean" : RED + "✗ Issues"), RESET);
            }
        }

        System.out.println("\n" + BOLD + " ▸ 3. DUPLICATE CHECK" + RESET);
        Set<String> uniqueCheck = new HashSet<>();
        long dups = 0;
        for (DataRecord r : records) { if (!uniqueCheck.add(r.toString())) dups++; }
        System.out.println(" Duplicate Rows: " + (dups > 0 ? RED : GREEN) + dups + RESET);

        System.out.println("\n" + BOLD + " ▸ 4. DATE VALIDATION" + RESET);
        if (colMap.containsKey("release_date")) {
            int dIdx = colMap.get("release_date");
            long invalid = 0;
            for(DataRecord r : records) {
                String val = r.getField(dIdx);
                boolean valid = false;
                if (!val.isEmpty()) {
                    for (String fmt : DATE_FORMATS) {
                        try {
                            SimpleDateFormat sdf = new SimpleDateFormat(fmt);
                            sdf.setLenient(false);
                            sdf.parse(val);
                            valid = true; break;
                        } catch (Exception e) {}
                    }
                }
                if(!valid) invalid++;
            }
            System.out.println(" Invalid Dates: " + (invalid == 0 ? GREEN : RED) + invalid + RESET);
        }

        System.out.println("\n" + BOLD + CYAN + "══════════════════════════════════════");
        System.out.println(" END OF REPORT");
        System.out.println("══════════════════════════════════════" + RESET);
    }

    private static String[] parseCsvLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        boolean inQuotes = false;
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '\"') inQuotes = !inQuotes;
            else if (c == ',' && !inQuotes) { fields.add(sb.toString()); sb.setLength(0); }
            else sb.append(c);
        }
        fields.add(sb.toString());
        return fields.toArray(new String[0]);
    }
}